File: src/main/java/com/jcloisterzone/feature/FishHut.java
package com.jcloisterzone.feature;

import com.jcloisterzone.board.Location;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.Rotation;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.event.ExprItem;
import com.jcloisterzone.event.PointsExpression;
import com.jcloisterzone.feature.River;
import com.jcloisterzone.game.capability.FishHutsCapability;
import com.jcloisterzone.game.capability.trait.WagonEligible;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTile;

import io.vavr.Tuple2;
import io.vavr.collection.HashSet;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import io.vavr.collection.Stream;

public class FishHut extends TileFeature implements WagonEligible, Completable, Monastic {
    private static final long serialVersionUID = 1L;

    protected final Set<FeaturePointer> neighboring; //for wagon move

    public FishHut(Location loc) {
        this(List.of(new FeaturePointer(Position.ZERO, FishHut.class, loc)), HashSet.empty());
    }

    public FishHut(List<FeaturePointer> places, Set<FeaturePointer> neighboring) {
        super(places);
        this.neighboring = neighboring;
    }

    @Override
    public FishHut setNeighboring(Set<FeaturePointer> neighboring) {
        if (this.neighboring == neighboring) return this;
        return new FishHut(places, neighboring);
    }

    @Override
    public Set<FeaturePointer> getNeighboring() {
        return neighboring;
    }

    @Override
    public Feature placeOnBoard(Position pos, Rotation rot) {
        return new FishHut(
            placeOnBoardPlaces(pos, rot),
            placeOnBoardNeighboring(pos, rot)
        );
    }

    @Override
    public boolean isOpen(GameState state) {
        return state.getDiagonalTiles2(getPosition()).size() < 4;
    }

    @Override
    public PointsExpression getStructurePoints(GameState state, boolean completed) {
        if (state.getCapabilities().get(FishHutsCapability.class) != null) {
        	return null;
        }

    	if (!completed) {
    		return null;
    	}

    	List<ExprItem> exprItems =  List.of(new ExprItem("fish-hut", 7));

        Position pos = places.get().getPosition();
        int riverTiles = state
        	.getDiagonalTiles(pos)
            .filter(pt -> state.getFeatureMap().get(pt.getPosition()).get().keySet().find(fp -> fp.getFeature().equals(River.class)).isDefined()).length();
        if (riverTiles>0) {
            exprItems = exprItems.append(new ExprItem(riverTiles, "river-tiles", riverTiles * 2));
        }
        return new PointsExpression("fish-hut", exprItems);
    }

    public Stream<PlacedTile> getRangeTiles(GameState state) {
        return state.getDiagonalTiles2(getPosition()).map(Tuple2::_2);
    }

    public static String name() {
        return "Yaga's Hut";
    }
}

File: src/main/ts/com/jcloisterzone/feature/FishHut.ts
import { HashSet, type Set } from "../../../io/vavr/Set.js";
import { List, Stream } from "../../../io/vavr/SeqTypes.js";
import { type Map as VMap } from "../../../io/vavr/Map.js";
import { Location } from "../board/Location.js";
import { Position } from "../board/Position.js";
import type { Rotation } from "../board/Rotation.js";
import { FeaturePointer } from "../board/pointer/FeaturePointer.js";
import { ExprItem } from "../event/ExprItem.js";
import { PointsExpression } from "../event/PointsExpression.js";
import type { ClassToken } from "../../../lang/Class.js";
import type { Capability } from "../game/Capability.js";
import { FishHutsCapability } from "../game/capability/FishHutsCapability.js";
import { LittleBuildingsCapability } from "../game/capability/LittleBuildingsCapability.js";
import type { WagonEligible } from "../game/capability/trait/WagonEligible.js";
import type { GameState } from "../game/state/GameState.js";
import type { PlacedTile } from "../game/state/PlacedTile.js";
import type { Completable } from "./Completable.js";
import type { Feature } from "./Feature.js";
import { MonasticFeature } from "./MonasticFeature.js";
import { River } from "./River.js";

/**
 * Fish hut (Fish Huts fan expansion). A monastic-style feature completed by its four
 * DIAGONAL tiles only: scores 7 when completed, plus 2 per diagonal river tile
 * (0 when incomplete). Deployable only with FishHutsCapability.
 */
export class FishHut extends MonasticFeature implements WagonEligible {
  isWagonEligible(): true {
    return true;
  }

  static readonly simpleName = "FishHut";

  constructor(loc: Location);
  constructor(places: List<FeaturePointer>, neighboring: Set<FeaturePointer>);
  constructor(a: Location | List<FeaturePointer>, neighboring?: Set<FeaturePointer>) {
    if (a instanceof Location) {
      super(List.of(new FeaturePointer(Position.ZERO, FishHut, a)), HashSet.empty<FeaturePointer>());
    } else {
      super(a, neighboring ?? HashSet.empty<FeaturePointer>());
    }
  }

  override setNeighboring(neighboring: Set<FeaturePointer>): Completable {
    if (this.neighboring === neighboring) return this as unknown as Completable;
    return new FishHut(this.getPlaces(), neighboring) as unknown as Completable;
  }

  placeOnBoard(pos: Position, rot: Rotation): Feature {
    return new FishHut(this.placeOnBoardPlaces(pos, rot), this.placeOnBoardNeighboring(pos, rot));
  }

  override isOpen(state: GameState): boolean {
    return state.getDiagonalTiles2(this.getPosition()).size() < 4;
  }

  override getRequiredCapability(): ClassToken<Capability<unknown>> {
    return FishHutsCapability as unknown as ClassToken<Capability<unknown>>;
  }

  getStructurePoints(state: GameState, completed: boolean): PointsExpression {
    let items = List.of(new ExprItem("fishhut", completed ? 7 : 0));
    if (completed) {
      const pos = this.getPlaces().head().getPosition();
      const riverTiles = state
        .getDiagonalTiles(pos)
        .filter((pt) =>
          state
            .getFeatureMap()
            .get(pt.getPosition())
            .get()
            .keySet()
            .find((fp) => fp.getFeature() === (River as never))
            .isDefined(),
        )
        .length();
      if (riverTiles > 0) {
        items = items.append(new ExprItem(riverTiles, "river-tiles", riverTiles * 2)) as List<ExprItem>;
      }
    }
    return new PointsExpression("fishhut", items);
  }

  override getRangeTiles(state: GameState): Stream<PlacedTile> {
    return state.getDiagonalTiles2(this.getPosition()).map((t) => t._2) as Stream<PlacedTile>;
  }

  /** Java Monastic default filters little buildings by getRangeTiles — for a fish hut
   *  that is the DIAGONAL tiles only (the TS MonasticFeature base uses a 3×3 box). */
  override getLittleBuildingPoints(state: GameState): List<ExprItem> {
    const buildings = state.getCapabilityModel<
      VMap<Position, LittleBuildingsCapability.LittleBuilding>
    >(LittleBuildingsCapability);
    if (buildings === null) {
      return List.empty<ExprItem>();
    }
    let tilePositions: Set<Position> = HashSet.empty<Position>();
    for (const pt of this.getRangeTiles(state)) tilePositions = tilePositions.add(pt.getPosition());
    const buildingsSeq = buildings.filterKeys((pos) => tilePositions.contains(pos)).values();
    return LittleBuildingsCapability.getBuildingsPoints(state, buildingsSeq);
  }
}
